geoopt/optim/rwolfe.py

```diff
@@ -0,0 +1,280 @@
+            c1=c1,
+            c2=c2,
+            fallback_stepsize=fallback_stepsize,
+            amax=amax,
+            amin=amin,
+            stabilize=stabilize,
+        )
+        super(RiemannianLineSearch, self).__init__(
+            params, defaults, stabilize=stabilize
+        )
+        self._params = []
+        for group in self.param_groups:
+            self._params.extend(group["params"])
+        if len(self.param_groups) > 1:
+            warning_string = """Multiple parameter groups detected.
+            Line search parameters will be taken from first group.
+            """
+            warnings.warn(warning_string, UserWarning)
+        self.line_search_method = self.param_groups[0]["line_search_method"]
+        if self.line_search_method not in ("wolfe", "armijo"):
+            raise ValueError(
+                f"Unrecognized line search method '{self.line_search_method}'"
+            )
+        self.c1 = self.param_groups[0]["c1"]
+        self.c2 = self.param_groups[0]["c2"]
+        self.fallback_stepsize = self.param_groups[0]["fallback_stepsize"]
+        self.amax = self.param_groups[0]["amax"]
+        self.amin = self.param_groups[0]["amin"]
+        self.old_phi = None
+        self.step_size_history = []
+        self.closure = None
+        self._step_size_dic = dict()
+
+    def phi_(self, step_size):
+        """Compute the line search objective, and store its derivatives in the state."""
+
+        if step_size in self._step_size_dic:
+            return self._step_size_dic[step_size]
+
+        param_copy = [param.clone() for param in self._params]
+
+        for point in self._params:
+            state = self.state[point]
+            if "grad" not in state:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            grad = state["grad"]
+
+            # compute retract and transport in gradient direction
+            new_point, grad_retr = manifold.retr_transp(point, -step_size * grad, grad)
+            with torch.no_grad():
+                point.copy_(new_point)
+
+            state["grad_retr"] = grad_retr
+
+        # recompute loss at new point
+        phi = self.closure()
+
+        # Store new gradients in state
+        for point in self._params:
+            grad = point.grad
+            if grad is None:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+
+            state["new_grad"] = grad
+
+            state["der_phi"] = torch.sum(
```

Comment:
btw, what happens in case you have more than one parameter? In here you compute norms, etc, per parameter. But from the conversation, I felt like you need a global norm, global inner product

Reply:
The `state["new_grad"]` isn't used, but should be. If the step we take is the one suggested by the line search, then `state["new_grad"]` already contains the gradient of the loss function with respect to this parameter, and we can omit a call to closure(). I had this idea in mind, but forgot to implement it. This does assume that the user does not alter the parameters in any way between consecutive calls of `.step()`, and I'm not whether such an assumption is necessarily good. 

`state["der_phi"]` is used in the function `derphi_`, where we just sum over all of the paremeters' `state["der_phi"]`. This sum is the global inner product we need. So, right, in principle we do not need to store it like this and we can just compute and store the sum from the get-go, but I figured storing this might not be bad since it tells us the contribution of a particular parameter to the gradient of the line search objective, which could be a useful metric for the user. (though this should be documented).

Reply:
> This does assume that the user does not alter the parameters in any way between consecutive calls 

I believe this covers 99.9% usecases. This is also the assumption in all geoopt optimizers that have momentum support.